### PR TITLE
`General`: Add breadcrumb navigation for student routes

### PR DESCRIPTION
**Pull request by bassner**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Breadcrumb navigation is currently only available in course management and admin views.
This PR enables breadcrumbs navigation for ``/courses/**`` so students can benefit from this enhanced UX as well.

### Description

- Enabled breadcrumb navigation for these routes in general
- Added translation overrides for clashes with segement names in course management
- Added some behavioral overrides for special cases, especially for backlinking to  ``courses/X/exercises`` from ``courses/X/[programming, modeling ..]-exercises/X/**``
- Removed "back button" from exercise headers

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Course with a lot of exercises and literally everything that a student can access in a course

--> In all steps: Verify that useful breadcrumbs are visible on top of the page and that they link back to the correct pages.

1. Go the the course overview
2. Select a course
3. Try each tab in the course overview
4. Try an exercise of each type
5. For each exercise type, open the submission view, add a new submission
6. For programming exercise, open the code editor as well
7. Visit a student exam view and try exercises there

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| navbar.component.ts | 75% | 85% |

### Screenshots

Screenshot 2022-06-06 at 17 16 04

### CODE REVIEW

**Review by jbllhff — APPROVED** [review]
Tested on ts2, works fine, nice breadcrumbs!

**Review by joaquinmurcia — DISMISSED** [review]
Tested on TS1, works as expected

**Review by rhubinger — DISMISSED** [review]
Tested on TS2, breadcrumbs look good.

**Review by willergeorg — DISMISSED** [review]
Code seems to be reasonable to me. 👍

**Review by SimonPaul02 — DISMISSED** [review]
Tested on TS2, works as expected. Nice feature!

**Review by yuling-sun — APPROVED** [review]
Tested on TS2. Breadcrumbs are working as described.

File: src/test/javascript/spec/component/shared/navbar.component.spec.ts [hunk 1]
```diff
@@ -563,4 +563,25 @@ describe('NavbarComponent', () => {
...
+                expect(exerciseTitleStub).toHaveBeenCalledTimes(1);
+
+                expect(component.breadcrumbs).toHaveLength(4);
+                expect(component.breadcrumbs[0]).toMatchObject({ uri: '/courses/', label: 'artemisApp.course.home.title' });
+                expect(component.breadcrumbs[1]).toMatchObject({ uri: '/courses/1/', label: 'Test Course' });
+                expect(component.breadcrumbs[2]).toMatchObject({ uri: '/courses/1/exercises/', label: 'artemisApp.courseOverview.menu.exercises' });
+                expect(component.breadcrumbs[3]).toMatchObject({ uri: '/courses/1/exercises/2/', label: 'Test Exercise' });
+            },
+        );
+    });
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/navbar.component.spec.ts:586]: Thanks for using parameterized tests! :rocket:

**Review by Dominik-Weinzierl — DISMISSED** [review]
Code looks reasonable and good to me :D

File: src/main/java/de/tum/in/www1/artemis/service/EntityTitleCacheService.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        NearCacheConfig nearCacheConfig = new NearCacheConfig() //
+                .setName(HAZELCAST_TITLE_CACHE + "-local") //
+                .setInMemoryFormat(InMemoryFormat.OBJECT) //
+                .setSerializeKeys(true) //
+                .setInvalidateOnChange(true) //
+                .setTimeToLiveSeconds(0) //
+                .setMaxIdleSeconds(0) //
+                .setEvictionConfig(evictionConfig) //
+                .setCacheLocalEntries(true);
+        config.getMapConfig(HAZELCAST_TITLE_CACHE).setNearCacheConfig(nearCacheConfig);
```
Comment (Dominik-Weinzierl) [inline src/main/java/de/tum/in/www1/artemis/service/EntityTitleCacheService.java:123]: Please wrap those lines in ` // @formatter:off` and `// @formatter:on` and remove those slashes :+1:

**Review by Dominik-Weinzierl — CHANGES_REQUESTED** [review]
One minor suggestion - i will look at the code in detail later.

File: src/main/java/de/tum/in/www1/artemis/service/EntityTitleCacheService.java (hunk 2 above)
Comment (bassner) [reply src/main/java/de/tum/in/www1/artemis/service/EntityTitleCacheService.java]: See e8b34e5 :)

**Review by babyygemperor — DISMISSED** [review]
Tested on ts2, works as described

File: src/test/javascript/spec/component/shared/navbar.component.spec.ts [hunk 3]
```diff
@@ -113,13 +113,13 @@ describe('NavbarComponent', () => {
...
 
-    it('should not build breadcrumbs for students', () => {
+    it('should build breadcrumbs for students', () => {
         const testUrl = '/courses/1/exercises';
         router.setUrl(testUrl);
 
         fixture.detectChanges();
 
-        expect(component.breadcrumbs.length).toEqual(0);
+        expect(component.breadcrumbs.length).toEqual(3);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/navbar.component.spec.ts:122]:
```suggestion
        expect(component.breadcrumbs).toHaveLength(3);
```

File: src/test/javascript/spec/component/shared/navbar.component.spec.ts [hunk 4]
```diff
@@ -563,4 +563,25 @@ describe('NavbarComponent', () => {
...
+        it.each(['programming-exercises', 'modeling-exercises', 'text-exercises'])(
+            'should replace exercise types in URI with just "exercise" on backlinking breadcrumbs',
+            (exType: string) => {
+                const testUrl = `/courses/1/${exType}/2`;
+                router.setUrl(testUrl);
+
+                fixture.detectChanges();
+
+                expect(courseManagementStub).toHaveBeenCalledTimes(1);
+                expect(exerciseTitleStub).toHaveBeenCalledTimes(1);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/navbar.component.spec.ts:577]:
```suggestion
                expect(exerciseTitleStub).toHaveBeenCalledOnce();
```

File: src/test/javascript/spec/component/shared/navbar.component.spec.ts [hunk 5]
```diff
@@ -563,4 +563,25 @@ describe('NavbarComponent', () => {
...
+    describe('Special student route breadcrumb cases', () => {
+        it.each(['programming-exercises', 'modeling-exercises', 'text-exercises'])(
+            'should replace exercise types in URI with just "exercise" on backlinking breadcrumbs',
+            (exType: string) => {
+                const testUrl = `/courses/1/${exType}/2`;
+                router.setUrl(testUrl);
+
+                fixture.detectChanges();
+
+                expect(courseManagementStub).toHaveBeenCalledTimes(1);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/navbar.component.spec.ts:576]:
```suggestion
                expect(courseManagementStub).toHaveBeenCalledOnce();
```

File: src/main/webapp/app/shared/layouts/navbar/navbar.component.ts [hunk 6]
```diff
@@ -265,6 +265,19 @@ export class NavbarComponent implements OnInit, OnDestroy {
...
+        exams: 'artemisApp.courseOverview.menu.exams',
+        exercises: 'artemisApp.courseOverview.menu.exercises',
+        lectures: 'artemisApp.courseOverview.menu.lectures',
+        learning_goals: 'artemisApp.courseOverview.menu.learningGoals',
+        statistics: 'artemisApp.courseOverview.menu.statistics',
+        discussion: 'artemisApp.metis.discussion.label',
+        code_editor: 'artemisApp.editor.breadCrumbTitle',
+        participate: 'artemisApp.submission.detail.title',
+        live: 'artemisApp.submission.detail.title',
+        courses: 'artemisApp.course.home.title',
```
Comment (Dominik-Weinzierl) [inline src/main/webapp/app/shared/layouts/navbar/navbar.component.ts:278]: I am not really happy with the inconsistency here :thinking: but its up to you ^^

**Review by Dominik-Weinzierl — COMMENTED** [review]
Code looks reasonable  to me (some minor suggestions - i was really nit-picky - sorry) - but we should test it and check if we find some issues regarding the usage of the Cache (especially on multiple instances) :+1:

File: src/main/webapp/app/shared/layouts/navbar/navbar.component.ts (hunk 6 above)
Comment (bassner) [reply src/main/webapp/app/shared/layouts/navbar/navbar.component.ts:278]: We're just picking the terms we need from the pages we're linking to in a breadcrumb. I don't see a big problem here as this mostly avoids redundant declarations of the same names

Comment (Dominik-Weinzierl) [reply src/main/webapp/app/shared/layouts/navbar/navbar.component.ts:280]: > We're just picking the terms we need from the pages we're linking to in a breadcrumb. I don't see a big problem here as this mostly avoids redundant declarations of the same names

Yes seems reasonable :+1:

**Comment by bassner** [comment]
TL;DR for reviewers: Added caching mechanisms on both the client and the server. There should be almost no ``.../title`` REST calls anymore; please keep your dev tools open while testing to verify that. For those that test locally, you can test that repeated calls to fetch a title of the same entity will not result in multiple DB calls, and that the server side cache for the entity is wiped if you modify it  (aka change the title of a course, for example).

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ExerciseHintService.java [hunk 7]
```diff
@@ -255,4 +257,25 @@ private Set<ExerciseHint> getAvailableExerciseHintsForTask(Optional<Integer> sub
...
+     */
+    @Cacheable(cacheNames = "exerciseHintTitle", key = "#{exerciseId + '-' + exerciseHintId}", unless = "#result == null")
+    public String getExerciseHintTitle(Long exerciseId, Long exerciseHintId) {
+        final var hint = exerciseHintRepository.findByIdElseThrow(exerciseHintId);
+
+        if (hint.getExercise() == null || !hint.getExercise().getId().equals(exerciseId)) {
+            throw new ConflictException("An exercise hint can only be retrieved if the exerciseIds match.", "exerciseHint", "exerciseIdsMismatch");
+        }
+
+        return hint.getTitle() == null ? null : hint.getTitle();
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ExerciseHintService.java:279]:
```suggestion
        return hint.getTitle();
```
I don't see any semantical difference to my suggestion

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Code looks good apart from one minor detail

**Comment by krusche** [comment]
Can we avoid `@CacheEvict` somehow? It would be better if we could tell Hibernate to automatically evict a cache when the object was changed, no matter in which part of the application this is done.
If at all, we should try to add it into the repository or service layer and not into the resource layer.

Another important aspect: Please make sure breadcrumbs are not shown when students participate in the exam mode. The exam mode should be as clean as possible and everything that disturbs or that allows you to accidentally navigate to a different page should be hidden.

**Review by Marvin-Vitcu — APPROVED** [review]
Tested on ts1. Works as intended.

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ExerciseHintService.java (hunk 7 above)
Comment (bassner) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/ExerciseHintService.java]: Thanks, solved in 3e828ca 👍

File: src/main/webapp/app/shared/layouts/navbar/navbar.component.ts [hunk 8]
```diff
@@ -268,18 +270,38 @@ export class NavbarComponent implements OnInit, OnDestroy {
         plagiarism_cases: 'artemisApp.plagiarism.cases.pageTitle',
     };
 
+    studentPathBreadcrumbTranslations = {
+        exams: 'artemisApp.courseOverview.menu.exams',
+        exercises: 'artemisApp.courseOverview.menu.exercises',
+        lectures: 'artemisApp.courseOverview.menu.lectures',
+        learning_goals: 'artemisApp.courseOverview.menu.learningGoals',
+        statistics: 'artemisApp.courseOverview.menu.statistics',
+        discussion: 'artemisApp.metis.discussion.label',
```
Comment (mallgaier) [inline src/main/webapp/app/shared/layouts/navbar/navbar.component.ts:279]:
```suggestion
        discussion: 'artemisApp.metis.communication.label',
```
`discussion` is outdated

**Review by mallgaier — CHANGES_REQUESTED** [review]
One very small change:

File: src/main/webapp/app/shared/layouts/navbar/navbar.component.ts (hunk 8 above)
Comment (bassner) [reply src/main/webapp/app/shared/layouts/navbar/navbar.component.ts]: Thanks, done in 4abeaa1

**Review by Dominik-Weinzierl — APPROVED** [review]
Works as expected during the testing session :+1:

**Review by Dominik-Weinzierl — APPROVED** [review]
Code looks also good :+1:

**Review by ge65cer — APPROVED** [review]
Tested in the testing session, did not find any issues.

**Review by mallgaier — APPROVED** [review]
Tested during Testing session. Issue is resolved; excellent work :
